Source/DafnyStandardLibraries/src/Std/Frames.dfy

```diff
@@ -0,0 +1,88 @@
+module Std.Frames {
+
+  import opened Termination
+
+  // A trait for objects with a Valid() predicate. Necessary in order to
+  // generalize some proofs, but also useful for reducing the boilerplate
+  // that most such objects need to include.
+  trait Validatable extends object {
+    // Ghost state tracking the common set of objects most
+    // methods need to read.
+    //
+    // Note this is used as the decreases clause for Valid(),
+    // but it is not necessarily the best choice for a decreases clause
+    // for other methods on implementing types.
+    // In particular, methods with loops that allocate new objects
+    // probably can't use Repr, because even if constructors
+    // ensure that the new object's Repr is fresh,
+    // it is not possible to prove the actual proof obligation,
+    // which is `old(Repr) decreases to (new object).Repr`.
+    // See FramesExamples.dfy for a detailed example.
+    //
+    // TODO: Okay it's now or never - Repr or repr?? :)
+    ghost var Repr: set<object>
+
+    // TODO: Similarly, would Invariant() be better?
+    ghost predicate Valid()
```

Comment:
I vote for `Valid`. The intention is that the condition be invariant, but it describes what it means for the data structure to be valid. Also, it's a bit funny to say that a method "requires and ensures the invariant", since it's the fact that the method "requires and ensures `X`" that makes `X` an invariant in the first place.